### PR TITLE
Migrate mqtt code

### PR DESCRIPTION
**Pull request by facundo-hm**
This PR is based on the code of https://github.com/ionia-corporation/ReactNativeTS/pull/11

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -49,6 +51,7 @@
     /* Experimental Options */
     // "experimentalDecorators": true,        /* Enables experimental support for ES7 decorators. */
     // "emitDecoratorMetadata": true,         /* Enables experimental support for emitting type metadata for decorators. */
+    "allowUnreachableCode": true
```
Comment (PCaponetti) [inline tsconfig.json]: why did you add this?

File: tsconfig.json [hunk 2]
```diff
@@ -33,12 +33,14 @@
...
-    // "paths": {},                           /* A series of entries which re-map imports to lookup locations relative to the 'baseUrl'. */
+    "baseUrl": "./",                       /* Base directory to resolve non-absolute module names. */
+    "paths": {
+      "mqtt/*": ["node_modules/mqtt/*"]
+    },                           /* A series of entries which re-map imports to lookup locations relative to the 'baseUrl'. */
     // "rootDirs": [],                        /* List of root folders whose combined content represents the structure of the project at runtime. */
     // "typeRoots": [],                       /* List of folders to include type definitions from. */
     // "types": [],                           /* Type declaration files to be included in compilation. */
-    "allowSyntheticDefaultImports": true      /* Allow default imports from modules with no default export. This does not affect code emit, just typechecking. */
+    "allowSyntheticDefaultImports": true,      /* Allow default imports from modules with no default export. This does not affect code emit, just typechecking. */
```
Comment (PCaponetti) [inline tsconfig.json]: also, this seems like a funny thing to enable

File: package.json [hunk 3]
```diff
@@ -7,10 +7,12 @@
...
     "ios": "node node_modules/react-native/local-cli/cli.js run-ios",
     "prebuild": "node node_modules/.bin/rimraf build",
-    "build": "tsc",
+    "build:ts": "tsc",
+    "build": "tsc && cp -r ./images ./build/images",
     "pretest": "npm run build",
     "test": "jest",
-    "postinstall": "node_modules/.bin/rn-nodeify --install crypto --hack"
+    "postinstall": "node_modules/.bin/rn-nodeify --install crypto --hack && npm run build:mqttClient",
+    "build:mqttClient": "webpack node_modules/mqtt/mqtt.js node_modules/mqtt/mqttClient.js --output-library mqtt --output-library-target commonjs2"
```
Comment (PCaponetti) [inline package.json:15]: 👍

File: tsconfig.json [hunk 4]
```diff
@@ -11,7 +11,7 @@
     "sourceMap": true,                        /* Generates corresponding '.map' file. */
     // "outFile": "./",                       /* Concatenate and emit output to single file. */
     "outDir": "./build",                      /* Redirect output structure to the directory. */
-    "rootDir": "./src",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
+    // "rootDir": "./src",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
```
Comment (PCaponetti) [inline tsconfig.json:14]: this is odd to me that we build out a src directory, but I'm seeing that that is a common directory to have in a final product of a react native solution

File: src/store/mqtt/actions.ts [hunk 5]
```diff
@@ -39,81 +51,113 @@ export function connect() {
...
   };
 }
 
 export function disconnect() {
   return async function(dispatch: Dispatch, getState: GetState) {
+    dispatch({ type: constants.DISCONNECT_START });
     const state = getState();
-    if (!state.mqtt.connected) {
-      console.log(`MQTT ERROR: attempting to disconnect while already disconnected`);
+    if (!client || !state.mqtt.connected) {
```
Comment (PCaponetti) [inline src/store/mqtt/actions.ts]: would love to switch to the actual mqtt client's understanding of whether or not we are connected, but I put that in an issue on the original repo: https://github.com/ionia-corporation/xively-react-boilerplate/issues

File: src/store/mqtt/actions.ts [hunk 6]
```diff
@@ -39,81 +51,113 @@ export function connect() {
...
+      throw err;
+    }
+
+    dispatch({ type: constants.CONNECTED });
+
+    client.on('error', (err: any) => dispatch(setError(`Parsing error: ${err}`)));
+    client.on('close', () => dispatch({ type: constants.CLOSE }));
+    client.on('offline', () => dispatch({ type: constants.OFFLINE }));
+    client.on('reconnect', () => dispatch(reconnect()));
+    client.on('message', (topic: string, message: MqttMessage) =>
```
Comment (PCaponetti) [inline src/store/mqtt/actions.ts]: would also love to (at the application level) do what was noted in the issues section on MQTT.js's repo: ```In addition, we listen on AppState.addEventListener("change", etc...) and invoke mqttClient._checkPing() if the state is active in order to force the connection to attempt to reconnect when the app comes into focus because the JS timers don't work while it's in the background.```

**Review by PCaponetti — CHANGES_REQUESTED** [review]

File: tsconfig.json (hunk 4 above)
Comment (facundo-hm) [reply tsconfig.json:14]: I had to comment that option because now we are importing a module from `node_modules/mqtt/` and ts throws an error when trying to load a file from outside of `./src` anyway the folder is referenced in the `include` prop.

File: tsconfig.json (hunk 1 above)
Comment (facundo-hm) [reply tsconfig.json]: In some moment webpack was generating some unreachable code like `if (true) {} else {}` and I set this option to avoid the error on the ts compilation but now seems that we can get rid of it.

File: src/store/mqtt/actions.ts [hunk 7]
```diff
@@ -133,7 +135,7 @@ export function disconnect() {
   return async function(dispatch: Dispatch, getState: GetState) {
     dispatch({ type: constants.DISCONNECT_START });
     const state = getState();
-    if (!client || !state.mqtt.connected) {
+    if (!client && !client.connected) {
```
Comment (PCaponetti) [inline src/store/mqtt/actions.ts]: this `&&` should be a `||` right?

**Review by PCaponetti — COMMENTED** [review]
is it possible to use client.connected everywhere and remove `connected` from the `MqttState` interface?

File: src/store/mqtt/actions.ts (hunk 7 above)
Comment (facundo-hm) [reply src/store/mqtt/actions.ts]: I think it's going to throw an error if client isn't defined and tries to access to the connected prop.

Comment (PCaponetti) [reply src/store/mqtt/actions.ts]: nope, javascript "short circuits" https://stackoverflow.com/questions/12554578/does-javascript-have-short-circuit-evaluation
